src/Adapters/Push/ParsePushAdapter.js

```diff
@@ -0,0 +1,147 @@
+"use strict";
+// ParsePushAdapter is the default implementation of
+// PushAdapter, it uses GCM for android push and APNS
+// for ios push.
+
+const Parse = require('parse/node').Parse;
+const GCM = require('../../GCM');
+const APNS = require('../../APNS');
+
+function ParsePushAdapter(pushConfig) {
+  this.validPushTypes = ['ios', 'android'];
+  this.senders = {};
+
+  // Initialize senders
+  for (let validPushType of this.validPushTypes) {
+    this.senders[validPushType] = [];
+  }
+
+  pushConfig = pushConfig || {};
+  let pushTypes = Object.keys(pushConfig);
+  for (let pushType of pushTypes) {
+    if (this.validPushTypes.indexOf(pushType) < 0) {
+      throw new Parse.Error(Parse.Error.PUSH_MISCONFIGURED,
+                            'Push to ' + pushTypes + ' is not supported');
+    }
+
+    let typePushConfig = pushConfig[pushType];
+    let senderArgs = [];
+    // Since for ios, there maybe multiple cert/key pairs,
+    // typePushConfig can be an array.
+    if (Array.isArray(typePushConfig)) {
+      senderArgs = senderArgs.concat(typePushConfig);
+    } else if (typeof typePushConfig === 'object') {
+      senderArgs.push(typePushConfig);
+    } else {
+      throw new Parse.Error(Parse.Error.PUSH_MISCONFIGURED,
+                            'Push Configuration is invalid');
+    }
+    for (let senderArg of senderArgs) {
+      let sender;
+      switch (pushType) {
+        case 'ios':
+          sender = new APNS(senderArg);
+          break;
+        case 'android':
+          sender = new GCM(senderArg);
+          break;
+      }
+      this.senders[pushType].push(sender);
+    }
+  }
+}
+
+/**
+ * Get an array of push senders based on the push type.
+ * @param {String} The push type
+ * @returns {Array|Undefined} An array of push senders
+ */
+ParsePushAdapter.prototype.getPushSenders = function(pushType) {
+  if (!this.senders[pushType]) {
+    console.log('No push sender for push type %s', pushType);
+    return [];
+  }
+  return this.senders[pushType];
+}
+
+/**
+ * Get an array of valid push types.
+ * @returns {Array} An array of valid push types
+ */
+ParsePushAdapter.prototype.getValidPushTypes = function() {
+  return this.validPushTypes;
+}
+
+ParsePushAdapter.prototype.send = function(data, installations) {
+  let deviceMap = classifyInstallation(installations, this.validPushTypes);
+  let sendPromises = [];
+  for (let pushType in deviceMap) {
+    let senders = this.getPushSenders(pushType);
+    // Since ios have dev/prod cert, a push type may have multiple senders
+    for (let sender of senders) {
+      let devices = deviceMap[pushType];
+      if (!sender || devices.length == 0) {
+        continue;
+      }
+      // For android, we can only have 1000 recepients per send
+      let chunkDevices = sliceDevices(pushType, devices, GCM.GCMRegistrationTokensMax);
+      for (let chunkDevice of chunkDevices) {
+        sendPromises.push(sender.send(data, chunkDevice));
```

Comment:
This is probably fine for a start but it's not really the right sending strategy for APNs.

What this will do for an active app is it will basically spam the dev cert with a bunch of invalid device tokens. Especially in the old APNs binary protocol (which we are still using in this PR, because it looks like there isn't an APNs module that uses the new HTTP/2 protocol), sending a push to an invalid token is extremely expensive. Sending a push to an invalid token will cause Apple to close the socket, and then you have to reopen the socket, which requires the server to validate the client's cert and the client to validate the server's cert (since APNs uses the TLS client auth option). That handshake process takes ~250ms. So in essence, if you sent a push to 100k invalid device tokens, it would take 100k * 250 ms = 25000 seconds to get through those 100k pushes, and Apple would probably block your server's IP from connecting to gateway.sandbox.push.apple.com because it would interpret this as a DoS attack. (This actually happens--we have periodically dealt with this server side when we pushed out buggy deploys in the past.)

What you need to is to examine each installation object and select the right set of certs for that installation on an installation-by-installation basis. The process should look something like this:

1. Check to see if the installation has an `appIdentifier` field (which is the bundle id of the app). If it's missing this field, go to step 4.
2. Find all certs that are associated with this bundle id. Hopefully, there's only one cert associated with that bundle id (this happens if the dev is smart enough to give the dev and prod versions of their app different bundle ids). In that case, only send the push using that cert.
3. If there are multiple certs for a given bundle id, then first send the push using a prod cert (as the prod cert is likely to be associated with more devices). If that fails, only then send the push using the dev cert.
4. If the installation has no `appIdentifier` field, then you just have to try sending the push with all certs. Try prod certs first, then dev certs.

In particular, we should definitely not be sending pushes using the prod cert and dev cert at the same time. Only send a push to a dev cert if sending the push using the prod cert fails.